Bottom: 76e16f6a38440ca020ba2701f8a1166c744856b8
Top:    8955bbfb2f4525662417924e52075445c50ea9fa
Author: Jason Wang <jasowang@redhat.com>
Date:   2017-03-01 13:39:55 +0800

Refresh of bypass

---

diff --git a/drivers/net/tun.c b/drivers/net/tun.c
index 30863e3..5875c93 100644
--- a/drivers/net/tun.c
+++ b/drivers/net/tun.c
@@ -2614,6 +2614,18 @@ struct socket *tun_get_socket(struct file *file)
 }
 EXPORT_SYMBOL_GPL(tun_get_socket);
 
+struct socket *tap_get_skb_array(struct file *file)
+{
+	struct tun_file *tfile;
+	if (file->f_op != &tap_fops)
+		return ERR_PTR(-EINVAL);
+	tfile = file->private_data;
+	if (!tfile)
+		return ERR_PTR(-EBADFD);
+	return &tfile->tx_array;
+}
+EXPORT_SYMBOL_GPL(tap_get_skb_array);
+
 module_init(tun_init);
 module_exit(tun_cleanup);
 MODULE_DESCRIPTION(DRV_DESCRIPTION);
diff --git a/drivers/vhost/net.c b/drivers/vhost/net.c
index 62ebe93..92ada6f 100644
--- a/drivers/vhost/net.c
+++ b/drivers/vhost/net.c
@@ -980,6 +980,32 @@ static struct socket *get_tap_socket(int fd)
 	return sock;
 }
 
+static struct socket *get_tap_skb_array(int fd)
+{
+	struct file *file = fget(fd);
+	struct skb_array *array;
+
+	if (!file)
+		return ERR_PTR(-EBADF);
+	array = tap_get_skb_array(file);
+	if (IS_ERR(array))
+		fput(file);
+	return array;
+}
+
+static struct socket *get_skb_array(int fd)
+{
+	struct skb_array *array;
+
+	/* special case to disable backend */
+	if (fd == -1)
+		return NULL;
+	array = get_tap_skb_array(fd)
+	if (!IS_ERR(array))
+		return array;
+	return ERR_PTR(-ENOTSOCK);
+}
+
 static struct socket *get_socket(int fd)
 {
 	struct socket *sock;
